internal/p2p/pqueue.go

```diff
@@ -247,21 +248,24 @@ func (s *pqScheduler) process() {
 
 			for s.pq.Len() > 0 {
 				pqEnv = heap.Pop(s.pq).(*pqEnvelope)
-				s.size -= pqEnv.size
-
-				// deduct the Envelope size from all the relevant cumulative sizes
-				for i := 0; i < len(s.chDescs) && pqEnv.priority <= uint(s.chDescs[i].Priority); i++ {
-					s.sizes[uint(s.chDescs[i].Priority)] -= pqEnv.size
-				}
-
-				s.metrics.PeerSendBytesTotal.With(
-					"chID", chIDStr,
-					"peer_id", string(pqEnv.envelope.To),
-					"message_type", s.metrics.ValueToMetricLabel(pqEnv.envelope.Message)).Add(float64(pqEnv.size))
 				select {
 				case s.dequeueCh <- pqEnv.envelope:
+					s.size -= pqEnv.size
+
+					// deduct the Envelope size from all the relevant cumulative sizes
+					for i := 0; i < len(s.chDescs) && pqEnv.priority <= uint(s.chDescs[i].Priority); i++ {
+						s.sizes[uint(s.chDescs[i].Priority)] -= pqEnv.size
+					}
+
+					s.metrics.PeerSendBytesTotal.With(
+						"chID", chIDStr,
+						"peer_id", string(pqEnv.envelope.To),
+						"message_type", s.metrics.ValueToMetricLabel(pqEnv.envelope.Message)).Add(float64(pqEnv.size))
 				case <-s.closer.Done():
 					return
+				default:
+					heap.Push(s.pq, pqEnv)
```

Comment:
I think the design of this queue is... not great, on the account of the fact that you need an incoming message to ever dequeue things, so I think making this non-blocking, is fine if the message rate is constant, but if there are bursty workloads, it could degrade in not great ways. 

Alternatives:
- we could decouple the incoming and outgoing messages, so that they're blocked (mostly) on waiting for incoming messages and the capacity of the outgoing channels. To do this, I think we'd need... more threads, but seems plausible? 
- if a peer slows down we let the back pressure force a disconnect earlier? I guess my working theory of the queues is that peers don't regularly back up and then recover? I think it's good to make the process less blocking, but is there a case where we'd ever hit this for a healthy peer?